janusgraph-core/src/main/java/org/janusgraph/diskstorage/util/StaticArrayEntryList.java

```diff
@@ -398,11 +401,89 @@ private static <E,D> EntryList of(Iterable<E> elements, StaticArrayEntry.GetColV
     private static <E,D> EntryList of(Iterator<E> elements, StaticArrayEntry.GetColVal<E,D> getter, StaticArrayEntry.DataHandler<D> dataHandler) {
         Preconditions.checkArgument(elements!=null && getter!=null && dataHandler!=null);
         if (!elements.hasNext()) return EMPTY_LIST;
+
+        EntryListComputationContext context = generateComputationContext();
+        applyElementsComputation(elements, getter, dataHandler, context);
+        return convert(context);
+    }
+
+    public static EntryListComputationContext generateComputationContext(){
         long[] limitAndValuePos = new long[10];
-        byte[] data = new byte[limitAndValuePos.length*15];
-        EntryMetaData[] metadataSchema = null;
-        int pos=0;
-        int offset=0;
+        return new EntryListComputationContext(limitAndValuePos, new byte[limitAndValuePos.length*15], null, 0, 0);
+    }
+
+    public static <E> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                           StaticArrayEntry.GetColVal<E,StaticBuffer> getter,
+                                           ExecutorService executorService) {
+        supplyEntryList(chunkedJobDefinition, getter, StaticArrayEntry.StaticBufferHandler.INSTANCE, executorService);
+    }
+
+    private static <E,D> void supplyEntryList(ChunkedJobDefinition<Iterator<E>, EntryListComputationContext, EntryList> chunkedJobDefinition,
+                                              StaticArrayEntry.GetColVal<E,D> getter,
+                                              StaticArrayEntry.DataHandler<D> dataHandler,
+                                              ExecutorService executorService){
+        assert chunkedJobDefinition !=null && getter!=null && dataHandler!=null;
+
+        executorService.execute(() -> {
+
+            Queue<Iterator<E>> chunksQueue = chunkedJobDefinition.getDataChunks();
+
+            if(!chunkedJobDefinition.getProcessingLock().tryLock()){
+                return;
```

Comment:
If it fails to retrieve the lock and returns, don't we fail to reserve any data without any error message/exception?

Reply:
It is expected scenario (even so hardly achievable). It's not exceptional scenario. 
The workflow is something like this:
1. We fetch pages one by one without waiting for those pages to be processed.
2. Anytime any page is received it adds it's data to the processing queue.
3. After the data is added to the processing queue it tries to enter this block which is protected by lock. In some situations this block execution could be slower then the speed of next pages retrieval (potentially but usually not practically). In such whenever some thread tries to enter this block but can't (due to another thread is working on data processing) this thread just drops without doing any work because we don't want to block waste this thread for non-computational job (i.e. block this thread).
4. As said in step `3` the thread which couldn't acquire the lock just drops. This could result in situation when previous thread already leaving the lock and the current thread dropped already. In such case, to not be left in situation when the data from the current thread isn't processed - we add another check after we released the lock. In case new pages arrived after we released the lock, the current processing thread goes back to step `3` and tries to enter the locked block again. And the process repeat. If we were lucky to enter the locked block - good, we need to process the data from any pages which are added into the queue. If we couldn't enter the lock - it's OK because it means that another thread is already processing this queue and it will do any necessary checks after it releases the lock.

With such technique pages retrieval doesn't necessary depend on pages deserialization. Potentially we could retrieve 2, 3, 10 pages while we are still working on the first page deserialization. 
That said, usually deserialization is much faster than pages retrieval. Thus, it may look like the process of pages fetching depends on pages deserialization, but in reality it's not.

Reply:
The only problem I saw with your current approach is, it would cause the order of results to be non-deterministic. Previously, the results from CQL backend are deterministic - e.g. if you call `g.V().toList()` multiple times, vertices will always show up in the same order.

Is it possible to avoid this breaking change? I would assume we need an extra field to mark the order of chunks and apply the results to the `EntryListComputationContext` accordingly.

Reply:
The order didn’t change. 
We request the new page only after we added the current page into the queue.
Thus, the order of result is the same as it is currently in master branch.

Reply:
The pages retrieval flow:
1. Fetch page and put its data into the queue.
2. Request the new page (async).
3. Try to start processing the queue (skip processing the queue if some other thread is processing it right now).

Reply:
The logic is simple: We can’t fetch next page without fetching previous page first. We don’t trigger the fetching of the next page until we added the result of the current page into the processing queue. Thus, it guarantees the proper order of the pages data in the processing queue. 
The only difference is that some other thread is processing its queue in parallel while we are fetching pages and adding pages data to the processing queue.
The processing thread removes data from the queue in FIFO order which guarantees that the order is deterministic (ordered from the first pages to the last page). 
Processing thread will never be able to process the next page’s data unless it processed the previous page’s data. This is enforced by the ‘dataChunks’ (ConcurrentLinkedQueue).

Reply:
Thanks, I missed the fact that chunks are fetched in FIFO order from the queue. There's no non-determinism then!